overlord/devicestate/devicemgr.go

```diff
@@ -2214,3 +2214,141 @@ func (m *DeviceManager) RemoveRecoveryKeys() error {
 
 	return secbootRemoveRecoveryKeys(recoveryKeyDevices)
 }
+
+// EncryptionSupportInfo describes what encryption is available and needed
+// for the current device
+type EncryptionSupportInfo struct {
+	// Disabled is set if to true encryption was furcefully
+	// disabled (e.g. via the seed pa partition), if set the rest
+	// of the struct content is not relevant.
+	Disabled bool
+
+	// StorageSafety describes the level safety properties
+	// requested by the model
+	StorageSafety asserts.StorageSafety
+	// Available is set to true if encryption is available on this device
+	// with the used gadget.
+	Available bool
+
+	// Type us set to the EncryptionType that can be used if
+	// Available is true.
+	Type secboot.EncryptionType
+
+	// UnvailableErr is set if the encryption support availability of
+	// the this device and used gadget do not match the
+	// storage safety requirements.
+	UnavailableErr error
+	// UnavailbleWarning describes why encryption support is not
+	// available in case it is optional.
+	UnavailableWarning string
+}
+
+var secbootCheckTPMKeySealingSupported = secboot.CheckTPMKeySealingSupported
+
+// checkEncryption verifies whether encryption should be used based on the
+// model grade and the availability of a TPM device or a fde-setup hook
+// in the kernel.
+func (m *DeviceManager) checkEncryption(st *state.State, deviceCtx snapstate.DeviceContext) (secboot.EncryptionType, error) {
+	model := deviceCtx.Model()
+
+	kernelInfo, err := snapstate.KernelInfo(st, deviceCtx)
+	if err != nil {
+		return "", fmt.Errorf("cannot check encryption support: %v", err)
+	}
+	gadgetSnapInfo, err := snapstate.GadgetInfo(st, deviceCtx)
+	if err != nil {
+		return "", err
+	}
+	gadgetInfo, err := gadget.ReadInfo(gadgetSnapInfo.MountDir(), nil)
+	if err != nil {
+		return "", err
+	}
+
+	res, err := m.encryptionSupportInfo(model, kernelInfo, gadgetInfo)
+	if err != nil {
+		return "", err
+	}
+	if res.UnavailableWarning != "" {
+		logger.Noticef("%s", res.UnavailableWarning)
+	}
+	// encryption disabled or preferred unencrypted: follow the model preferences here even if encryption would be available
+	if res.Disabled || model.StorageSafety() == asserts.StorageSafetyPreferUnencrypted {
+		res.Type = secboot.EncryptionTypeNone
+	}
+
+	return res.Type, res.UnavailableErr
+}
+
+func (m *DeviceManager) encryptionSupportInfo(model *asserts.Model, kernelInfo *snap.Info, gadgetInfo *gadget.Info) (EncryptionSupportInfo, error) {
+	secured := model.Grade() == asserts.ModelSecured
+	dangerous := model.Grade() == asserts.ModelDangerous
+	encrypted := model.StorageSafety() == asserts.StorageSafetyEncrypted
+
+	res := EncryptionSupportInfo{
+		StorageSafety: model.StorageSafety(),
+	}
+
+	// check if we should disable encryption non-secured devices
+	// TODO:UC20: this is not the final mechanism to bypass encryption
+	if dangerous && osutil.FileExists(filepath.Join(boot.InitramfsUbuntuSeedDir, ".force-unencrypted")) {
+		res.Disabled = true
```

Comment:
we could return here already?

Reply:
Happy to to do that. I was mostly thinking there is no harm in showing is encryption would be possible on the device with a different model but probably silly reasoning.

Reply:
I updated this now.